packages/react-query/src/__tests__/queryOptions.test-d.tsx

```diff
@@ -241,10 +243,40 @@ describe('queryOptions', () => {
     }
 
     const options = queryOptions({
-      queryKey: ['key'],
+      queryKey: ['key'] as QueryKey,
```

Comment:
why is this type assertion necessary? I don’t think this should be needed, because it means we can’t pass what `queryOptions` creates to a function that accepts `UseQueryOptions` 🤔

Reply:
Okay got to the bottom of this one

So now that queryOptions is correctly typed (it was wrong before) it knows the query key is a `string[]` - this is broadly a good thing

`UseQueryOptions<number>` though has the query key down as `QueryKey` by default, also broadly a good thing.

But `UseQueryOptions` also contains a callback `QueryFunction` which includes the query key

What happens when typescript structurally tries to treat a supertype as the subtype like it does when comparing types passed in callbacks? It says no "you can't assign supertype to subtype". The solution is:

1. type userland code correctly (see the updated test where I use `UseQueryOptions<number, any, any, Array<string>>` or could use `UseQueryOptions<number, any, any, any>`
2. Change some default query key types in the codebase to `any` and ensure that wherever react-query uses the types internally they get the parameters passed in (should be doing this anyway but this PR fixes a few cases of that being missed)

Reply:
yeah I think 2 is better. I think I’ve stumbled upon this before, too. can you do that?

Reply:
I can, though I am conscious of how this could inadvertently de-type folk's codebases because many/most types are able to be used both internally and in userland

Will have a look into it it

Reply:
Okay so I've made this switch on all interfaces and types, but not on any functions or classes and it's not looking like the right way

Inference in DataTags breaks, because of course `any extends DataTag<any, infer Data, any> ? true : false` is always `true` and there are other similar classes of issues. Test suite here is up to date and number of errors not that huge but I don't see a way around them right now

Reply:
Just to make a case, we're in a position already where some subtle breaking changes went out because of my previous PRs as I missed them. 

This PR with option 1 above, moves those breakages to somewhere that userland can trivially fix by typing interfaces correctly, and should not affect the vast majority of codebases anyway, whereas tRPC can't actually use 5.62.2 right now (for reasons I'd need to refresh my brain on but they're the reason I came back with this PR)

Reply:
I can't say it's better, took quite a few changes to fix query-core and now every package has problems - pushed at this point so you can at least review the classes of errors. Seems a much bigger impact for the public surface area than my original idea on this PR